presto-hive/src/main/java/com/facebook/presto/hive/NestedColumn.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+public class NestedColumn
```

Comment:
I don't think we need this class as we already have `Subfield`

Reply:
I initially used the Subfield class, but the there is already a member variable with list of needed Subfields in HiveColumnHandle. Rather than adding one more Subfield type member variable to represent the NESTED_COLUMN type definition, created this class which is looked at when the column type is NESTED_COLUMN in HiveColumnHandle. Also Subfield represents more than NestedColumn, it has index etc. Let me know if you think it is ok to use Subfield in this case.

Reply:
NESTED_COLUMN type doesn't fit into ColumnType. ColumnType is the attribute of that column instead of the structure. Why don't we merge the existing subfield together with the one you want in the optimizer rule? I assume you also need the subfield right?

Reply:
Currently the way Subfield in HiveColumnHandle is used is as a hint to the reader telling it to read only the required subfields from the complex column. The type (RowType) of the column from the reader doesn't change. 

For example `col` of type `RowType(int a, int b, int c)`. If only `col.a` is required the returned type is of column is still `RowType(int a, int b, int c)` but only the Block of `a` gets read and added to RowBlock. For the subcolumns `b` and `c`, RLEBlock is returned in RowBlock. 

In this patch we are changing the output type to be the type of the nested column. In the same above example: `col.a` is selected, the reader just returns the IntBlock and the column name is a derived name. For this reason we added the new ColumnType as the reader interprets this column differently. 

We can replace NestedColumn with Subfield, but the ColumnType(NESTED_COLUMN) is needed. May be rename the NESTED_COLUMN to SUBFIELD?

Reply:
ColumnType should not denote the if a column is structured or not. It is to denote the attribute. A column could be a regular/synthesized column but still be nested. On a side question, I don't see NESTED_COLUMN is used in this commit? Did I miss anything?

Reply:
@highker @zhenxiao  Here are the changes that we discussed offline.

- don't add any new `columnType` type in `HiveColumnHandle` -- removed it
- remove `NestedColumn` class and use `Subfield` instead - done
- piggyback on `List<Subfield> requiredSubfields` to store the pushed down Nested column - unfortunately this has issues. This particular field is always populated when there is any Subfield type such as `a[1].b`. This type of `Subfield`s are not handled in current Parquet reader. So we can't rely on this list being size 1 to convey to the Parquet reader that the Subfield in the list is a pushed down dereference. Instead of piggybacking on `List<Subfield> requiredSubfields` to store the pushed down, added a new field called `Optional<Subfield> pushdownSubfield`. This has no impact on existing readers.

Also added the remaining patches that implement the feature end-2-end including the pushdown of filters on dereference columns to Parquet reader. 

Let me know if there are any changes needed.